src/Server/TextDocument.php

```diff
@@ -148,4 +167,14 @@ public function formatting(TextDocumentIdentifier $textDocument, FormattingOptio
         return [$edit];
     }
     
+    public function completion(TextDocumentIdentifier $textDocument, Position $position)
+    {
+        $completionReporter = new CompletionReporter();
+        $tokens = token_get_all($this->documents[$textDocument->uri]);
```

Comment:
See above comment, this will essentially re-parse the whole document, even though PHPParser already did this. If PHPParser is really not capable of parsing nodes up until the cursor, we should take as many nodes as it can provide and then only use `token_get_all()` for the substring between the last node and the cursor, if `token_get_all` will even work for this. Might even have to use regexps for those last chars in front of the cursor.